Update contributing.md with gathered notes from top contributors on their expectations for new committers (#21682)
prestodb/presto: CONTRIBUTING.md

Presto welcomes contributions from everyone.

Contributions to Presto should be made in the form of GitHub pull request submissions and reviews.

Each pull request submission will be reviewed by a contributor or committer in the project.  In order for a PR to be eligible to be merged, a committer for the appropriate code must approve the code. Once approved by a committer, the PR may be merged by anyone.

Pull request reviews are encouraged for anyone in the community who would like to contribute to Presto, and are
expected from contributors and committers in at least equal proportion to their code contributions.

> Contributions should have an associated GitHub issue.

Reviewer: Not sure this is needed for each issues

> Contributions should have an associated GitHub issue.

Reviewer 2: Let's attempt to break this down more precisely using T-shirt sizes:

1) Large chances should have an RFC.  The RFC should be reviewed before patches are submitted.
2) Medium size changes should have an issue.  Work from RFCs may be broken down into smaller issues, which themselves link to the RFC.
3) Smaller changes, such as minor bug fixes and code formatting, may forego an issue and just submit a PR.

The sizing of each will require intuition which we will develop as a community.

> Contributions should have an associated GitHub issue.

Reviewer 3: Works for me!

> Contributions should have an associated GitHub issue.

Author: Integrated

## Minimum Expectations for Contributing to Presto
It is expected that everyone wishing to commit code will:
* Work through the “Getting Started” materials
* Read and agree to the Code of Conduct
* Join Presto Slack
* File an issue

> * Self verify new code matches codestyle for Presto

Author: Added

> * Follow commit standards

Author: Will get those added!

* Add or modify existing tests related to code changes being submitted
* Run and ensure local tests pass before submitting a merge request

## Designing Your Code
* Consider your code through 3 axes

>     1. Code Quality/Maintainability, e.g

Author: Agreed, will make the change

        1. Code Style
        2. Maintainability - is the feature added in a hacky way (if we did everything that way would the code base become hard to use),  or is it implemented in a clean way with appropriate interfaces, fits in with the rest of Presto's design etc. (so keeps our code base maintainable).
    1. Code Safety, e.g.
        1. Adequate testing
        1. Feature flags to enable/disable features that might be risky
        1. Thread safe

>         1. No data structures that can grow without bounds. memory usage is accounted for

Reviewer 2: ```suggestion
        1. No data structures that can grow without bounds. Memory usage is accounted for
```

I will create an issue for adding this to our functions doc--very important to get right, and it's missing!  CC: @ZacBlanco

>         1. No data structures that can grow without bounds. memory usage is accounted for

Author: Good catch

        1. Not introducing expensive calls in a performance sensitive area
    1. User friendliness e.g.
        1. Config options have names and descriptions that can be understood by someone configuring Presto
        1. All new language features, new functions, and major features have documentation added
        1. Release notes following release note guidelines are added for user visible changes

> * For large features, discuss your design with relevant code owners before you start implementing it.

Reviewer 2: Perhaps mention that large features may need an RFC: https://github.com/prestodb/rfcs

Author: added elsewhere with link to RFC

## Code Style

We recommend you use IntelliJ as your IDE. The code style template for the project can be found in the [codestyle](https://github.com/airlift/codestyle) repository along with our general programming and Java guidelines. In addition to those you should also adhere to the following:

* **Naming**
    * Avoid abbreviations
      E.g. positionCount instead of positionCnt
    * Line width and spacing
        * Lines should be within 180 chars. Intellij “Reformat code” does NOT enforce the line width, so you’ll need to adjust it yourself.
        * Long function declarations exceeding 180 characters shall be broken down into multiple lines, one argument/parameter per line. A good example is like:

      ```java
          public ParquetPageSource(
                  ParquetReader parquetReader,
                  List<Type> types,
                  List<Optional<Field>> fields,
                  List<String> columnNames,
                  RuntimeStats runtimeStats)
      
      ```

>         * Do not put the first parameter on the same line as the function name, or move the parentheses to different lines. Bad example:

Reviewer 3: Nice work providing an example of incorrectly broken down code for the reader to compare their code to. No change requested here, this is another callout of good work.

                            !(isIndexColumn && field.isPresent()),
                            "Field info for row index column must be empty Optional"));
        }
      ```

* **Ordering of class members and methods**
    * Class members are in front of methods.
    * Group and order the class members and methods in the order of their access levels in desc order:
        1. public
        1. protected
        1. package private
        1. private
    * Group and order the class members:
        1. static final
        1. final
        1. normal
    * Order the methods with the same access level in the order they’re called.
* **Encourage proper encapsulation**
    * Do not recommend having public class members. Use lower access levels as much as possible. Exceptions may apply for performance critical paths, in that case discuss your design first with relevant code owners.
    * Make a function access level as low as possible
    * If a public class is only used by one caller and the usage is very local, consider making it a nested class.
* **Immutable and thread safe classes**
    * Whenever possible, class fields should be final

>     * When it's not possible, ensure that accesses to the non-final fields are thread safe by whatever methods are appropriate for the circumstance (Concurrent collections, synchronized access, etc.)

Reviewer 2: ```suggestion
    * When it's not possible, ensure that accesses to the non-final fields are thread safe by whatever methods are appropriate for the circumstance (Concurrent collections, synchronized access, etc.) if they are in a code path that is multi-threaded
```

             .collect(Collectors.toList())
  ```
    
  * We use ```toImmutableList()``` e.g.
    
  ```java
  import com.google.common.collect.ImmutableList;
  ...
  import static com.google.common.collect.ImmutableList.toImmutableList;
  ...
  expressions.stream()
             .map(OriginalExpressionUtils::castToExpression)
             .collect(toImmutableList())
  ```

* **Method argument validation**
    * Usually needed for constructors

      ```java
      protected SqlScalarFunction(Signature signature)
      {
          this.signature = requireNonNull(signature, "signature is null");
          checkArgument(signature.getKind() == SCALAR, "function kind must be SCALAR");
      }
      ```

* **Use proper annotations**
    * @Nullable
    * @VisibleForTesting
    * @Override
    * @Experimental
* **Use of Optional instead of bare metal nullable objects when appropriate**
    * public method parameters
    * Performance NON-CRITICAL path
    * Creating Optional objects is not free. If the function call is in a critical loop, do NOT use Optional parameters.
* **Alphabetize**
    * Sections in documentation
    * Methods
    * Variables
    * Sections
* **Comment styles**

>     * Recommend to add the “/** */” (Javadoc) style comments to all interface methods, with explanation of the parameters and returns. E.g.

Reviewer 3: ```suggestion
    * Recommend to add the “/** */” (Javadoc) style comments to all interface methods, with explanation of the parameters and returns. For example:
```

               * @param elementOptional an element to be transformed. Will be empty
               * when there are no more elements. In such case transformation should
               * finish processing and flush any remaining data.
               * @return the current transformation state, optionally bearing a result
               * @see TransformationState#needsMoreData()
               * @see TransformationState#blocked(ListenableFuture)
               * @see TransformationState#yield()
               * @see TransformationState#ofResult(Object)
               * @see TransformationState#ofResult(Object, boolean)
               * @see TransformationState#finished()
               */
              TransformationState<R> process(Optional<T> elementOptional);
          }
      ```

    * Recommend to add the “/** */” style comments to important or difficult public methods, with explanation of the parameters and returns.
    * Within the method body, use “//” style comments on separate lines for difficult parts.
    * Use “//” style comments for class members if it can help understanding the code. Can be on the same line or separate lines in front.
* **Code succinctness**

>     * Inline function calls when appropriate. E.g. if a function is only called once, we don’t need to create a variable for it.

Reviewer 4: ```suggestion
    * Inline function calls when appropriate. For example, if a function is only called once, we don’t need to create a variable for it.
```

    * Lambda

>       E.g. Instead of doing this

Reviewer 4: ```suggestion
      For example, Instead of doing this
```

      ```java
      Iterables.getOnlyElement(argumentSets).stream()
      .map((rowExpression) -> OriginalExpressionUtils.castToExpression(rowExpression))
      ```
      We just do
      ```java
      Iterables.getOnlyElement(argumentSets).stream()
      .map(OriginalExpressionUtils::castToExpression)
      ```

* When appropriate use Java 8 Stream API
* Categorize errors when throwing an exception

## Commit Standards
* ### Commit Size
    * Recommended lines of code should be within +1000 lines, and should focus on one single major topic.\
      If your commit is more than 1000 lines, consider breaking it down into multiple commits, each handling a specific small topic.
* ### Commit Message Style
    * **Separate subject from body with a blank line**
    * **Subject**
        * Limit the subject line to 10 words or 50 characters
        * If you cannot make the subject short, you may be committing too many changes at once
        * Capitalize the subject line
        * Do not end the subject line with a period
        * Use the imperative mood in the subject line
    * **Body**
        * Wrap the body at 72 characters

>         * Use the body to explain what and why vs. how

Reviewer 3: ```suggestion
        * Use the body to explain what and why versus how
```
GitLab doc style guide doesn't mention vs. or versus. The [Google developer doc word list entry for vs.](https://developers.google.com/style/word-list#vs) recommends using the unabbreviated version.

        * Use the imperative mood in the body\

>           E.g. “If applied, this commit will ___________”

Reviewer 4: ```suggestion
          For example, “If applied, this commit will ___________”
```

        * Communicate only context (why) for the commit in the subject line

>         * Use the body for What and Why, if your commit is complex or dense, share some of the How context

Reviewer 3: ```suggestion
        * Use the body for What and Why, 
        * If your commit is complex or dense, share some of the How context in the body
```
This felt like it could be two separate guideline points: let me know if you feel strongly that this should stay one line.

        * Assume someone may need to revert your change during an emergency
    * **Content**
        * **Aim for smaller commits for easier review and simpler code maintenance**
        * All bug fixes and new features must have associated tests

>         * Commits should pass tests on their own, not be dependent on other commits\

Reviewer 4: ```suggestion
        * Commits should pass tests on their own, not be dependent on other commits
```

        * The following is recommended:
            * Describe why a change is being made.
            * How does it address the issue?
            * What effects does the patch have?
            * Do not assume the reviewer understands what the original problem was.

>             * Do not assume the code is self-evident/self-documenting.

Reviewer 3: ```suggestion
            * Do not assume the code is self-evident or self-documenting.
```
Avoid slashes. See [GitLab doc recommended word list entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes) for advice and suggestions.

            * Read the commit message to see if it hints at improved code structure.
            * The first commit line is the most important.
            * Describe any limitations of the current code.
            * Do not include patch set-specific comments.

Details for each point and good commit message examples can be found on https://wiki.openstack.org/wiki/GitCommitMessages#Information_in_commit_messages

* **Metadata**

>     * If the commit was to solve an Github issue, refer to it at the end of a commit message in a rfc822 header line format.\

Reviewer 3: ```suggestion
    * If the commit was to solve a Github issue, refer to it at the end of a commit message in a rfc822 header line format.\
```

>       E.g.\
>       Resolves: #1234\
>       See also: #1235, #1236\

Reviewer 4: ```suggestion
      For example,\
      Resolves: #1234\
      See also: #1235, #1236
```

* **Backport Commits**
    * Must refer to original commit ID (refer to the PR link in the PR message) or link and author in the commit message
      ```bash
      Fix OOM caused by foo in bar
      
      Foo was pack ratting ByteBuffers causing OOM.
      
      Cherry-pick of https://github.com/prestodb/presto/pull/18424/commits/30e36dae0cdb9debd991931370b86301cd17e261 
      Co-authored-by: Foo Bar <foo@bar.com>
      ```

>     * Try to follow the original PR/commits structure. Do not mingle multiple original commits or reorder them.

Reviewer 3: ```suggestion
    * Try to follow the original PR and commits structure. Do not mingle multiple original commits or reorder them.
```
Avoid slashes. See [GitLab doc recommended word list entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes) for advice and suggestions.

New committers are approved by majority vote of the TSC ([see TSC charter](https://github.com/prestodb/tsc/blob/master/CHARTER.md)).  To become a committer, reach out to an [existing TSC member](https://github.com/prestodb/tsc#members) and ask for their feedback on your eligibility (see: [How to become a Presto Committer?](https://github.com/prestodb/presto/wiki/How-to-become-a-Presto-committer%3F)).  Note: to expedite the process, consider creating a document that outlines your Github stats, such as the number of reviews, lines of code added, number of PRs, and outlines particularly outstanding code and review contributions.  If the TSC member believes you are eligible, they will submit your nomination to a vote by the TSC, typically in the form of a PR that adds your handle to the `CODEOWNERS` file.  The process is complete once the PR is merged.

## Pull Requests
* #### PR size and structure

>     * A PR can consist of multiple small commits, preferably not exceeding 20(?).

Reviewer 2: ```suggestion
    * A PR can consist of multiple small commits, preferably not exceeding 20.
```

We can keep it like this for now and adjust if needed

>     * The total number of lines modified in a single PR shall not exceeding 5000(?)

Reviewer 2: ```suggestion
    * The total number of lines modified in a single PR shall not exceeding 5000, except for changes that include checked in code generated files (such as [presto_protocol.cpp](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/presto_protocol/presto_protocol.cpp)).
```

    * The commits sequence shall be in the dependencies order, not in date created/modified order
    * Every commit in the PR shall pass all the tests
      Example: https://github.com/prestodb/presto/pull/12991/commits

* #### PR title

>     * Please refer to the commit title

Reviewer 4: It might be worth linking to the relevant section here or maybe just add the word "above" somewhere in this line

```suggestion
    * Please refer to the section on commit subject
```

* #### PR message
    * Tests
    * Release notes
    * Metadata

>         * Resolves #1234

Reviewer 2: I think we can simplify this and simply instruct people to follow the PR template, which has check boxes one must check.

Author: instructed them to follow the PR template provided

* #### Backport PR
    * Must refer to the original PR and provide link in the PR message
    * Must refer to original authors in the PR message
    * Do NOT change original commits titles or messages.
    * If original PR contains multiple commits, follow the same structure.

## Pull Request Checklist

> If you are not yet an official committer, you must fork from Master and then submit a Pull Request to contribute as an individual.

Reviewer 3: ```suggestion
If you are not an official committer, you must fork from Master and then submit a Pull Request to contribute as an individual.
```

> If you have commit permissions:

Reviewer 2: I think this can be simplified, since even committers must have a fork and then submit a PR from their fork.

It might make sense to link to the fork and pull model, since that is what we follow: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/about-collaborative-development-models#fork-and-pull-model

Author: added and adjusted some language

Presto welcomes contributions from everyone.

Contributions to Presto should be made in the form of GitHub pull request submissions and reviews.

Each pull request submission will be reviewed by a contributor or committer in the project.  In order for a PR to be eligible to be merged, a committer for the appropriate code must approve the code. Once approved by a committer, the PR may be merged by anyone.

Pull request reviews are encouraged for anyone in the community who would like to contribute to Presto, and are
expected from contributors and committers in at least equal proportion to their code contributions.

Contributions should have an associated GitHub issue.

## Minimum Expectations for Contributing to Presto
It is expected that everyone wishing to commit code will:
* Work through the “Getting Started” materials
* Read and agree to the Code of Conduct
* Join Presto Slack
* File an issue
* Self verify new code matches codestyle for Presto
* Follow commit standards

> * Add or modity existing tests related to code changes being submitted

Reviewer 5: ```suggestion
* Add or modify existing tests related to code changes being submitted
```

            this.rowIndexLocations = requireNonNull(rowIndexLocations, "rowIndexLocations is null");
            this.columnNames = ImmutableList.copyOf(requireNonNull(columnNames, "columnNames is null"));
            this.runtimeStats = requireNonNull(runtimeStats, "runtimeStats is null");
      
            checkArgument(
                    types.size() == rowIndexLocations.size() && types.size() == fields.size(),
                    "types, rowIndexLocations, and fields must correspond one-to-one-to-one");
      
            Streams.forEachPair(
                    rowIndexLocations.stream(),
                    fields.stream(),
                    (isIndexColumn, field) -> checkArgument(
                            !(isIndexColumn && field.isPresent()),
                            "Field info for row index column must be empty Optional"));
        }
      ```

* **Ordering of class members and methods**
    * Class members are in front of methods.
    * Group and order the class members and methods in the order of their access levels in desc order:
        1. public
        1. protected
        1. package private

>            iv.	private

Reviewer 5: 1. private ?

        1. final
        1. normal
    * Order the methods with the same access level in the order they’re called.
* **Encourage proper encapsulation**
    * Do not recommend having public class members. Use lower access levels as much as possible. Exceptions may apply for performance critical paths, in that case discuss your design first with relevant code owners.
    * Make a function access level as low as possible
    * If a public class is only used by one caller and the usage is very local, consider making it a nested class.
* **Immutable and thread safe classes**
    * Whenever possible, class fields should be final
    * When it's not possible, ensure that accesses to the non-final fields are thread safe by whatever methods are appropriate for the circumstance (Concurrent collections, synchronized access, etc.)

* **Static imports**
  E.g. in your code, do not use something like

    ```java
    String.format("%s", rule.getClass().getSimpleName())
    ```

    * Instead, static import this method and just use “format()”

        ```java
        import static java.lang.String.format;
        ...
        format("%s", rule.getClass().getSimpleName())
      
        ```

>         * Similarly, do NOT use

Reviewer 5: reduce one level of indentation. I guess this should be the same as line 246
the thing applies to line 262, 269, and 278

        ```java
        expressions.stream()
                   .map(OriginalExpressionUtils::castToExpression)
                   .collect(Collectors.toList())
      
        ```

        * We use ```toImmutableList()``` e.g.

        ```java
        import com.google.common.collect.ImmutableList;
        ...
        import static com.google.common.collect.ImmutableList.toImmutableList;
        ...
        expressions.stream()
                   .map(OriginalExpressionUtils::castToExpression)
                   .collect(toImmutableList())
        ```

* **Method argument validation**
    * Usually needed for constructors

      ```java
      protected SqlScalarFunction(Signature signature)
      {
          this.signature = requireNonNull(signature, "signature is null");
          checkArgument(signature.getKind() == SCALAR, "function kind must be SCALAR");
      }
      ```

* **Use proper annotations**
    * @Nullable
    * @VisibleForTesting
    * @Override
    * @Experimental
* **Use of Optional instead of bare metal nullable objects when appropriate**
    * public method parameters
    * Performance NON-CRITICAL path

>     * Creating Optional objects is not free. If the function call is in a critical loop, do NOT use Optional parameters. A

Reviewer 5: ```suggestion
    * Creating Optional objects is not free. If the function call is in a critical loop, do NOT use Optional parameters.
```

               * @see TransformationState#ofResult(Object)
               * @see TransformationState#ofResult(Object, boolean)
               * @see TransformationState#finished()
               */
              TransformationState<R> process(Optional<T> elementOptional);
          }
      ```

    * Recommend to add the “/** */” style comments to important or difficult public methods, with explanation of the parameters and returns.
    * Within the method body, use “//” style comments on separate lines for difficult parts.
    * Use “//” style comments for class members if it can help understanding the code. Can be on the same line or separate lines in front.
* **Code succinctness**
    * Inline function calls when appropriate. E.g. if a function is only called once, we don’t need to create a variable for it.
    * Lambda
      E.g. Instead of doing this

      ```java
      Iterables.getOnlyElement(argumentSets).stream()
      .map((rowExpression) -> OriginalExpressionUtils.castToExpression(rowExpression))
      ```
      We just do
      ```java
      Iterables.getOnlyElement(argumentSets).stream()
      .map(OriginalExpressionUtils::castToExpression)
      ```

> * When appropriate use Java 8(?) Stream API

Reviewer 5: yea, Java 8

    * Implement and address existing feedback before requesting further review
    * Make a good faith effort to locate example or referential code before requesting someone else direct you towards it
    * If you see a lack of documentation, create a separate commit with draft documentation to fill the gap
        * This documentation can be iterated on same as any code commit, demonstrate in real time that you are learning the code section
    * Implement or modify relevant tests, otherwise provide clear explanation why test updates were not necessary
    * Tag your PR with affected code areas as best as you can, it’s okay to tag too many, better to cut down irrelevant tags than miss getting input from relevant subject matter experts

- Example PR message
    - please refer [here](https://github.com/prestodb/presto/blob/master/pull_request_template.md)

### What not to do for Pull Requests
* Submit before getting peer review in your own organization
* Request review without addressing/implementing previous feedback
* Asking reviewers to provide examples or code references instead of finding them on your own
* Protesting lack of documentation for a code section

>     * Insread, review the related code, draft initial documentation as a separate commit

Reviewer 3: ```suggestion
    * Instead, review the related code, then draft initial documentation as a separate commit
```
Typo and light revision

* Submit without test cases or clear justification for lack thereof

## Code Reviews
#### What to do
* Provide explicit feedback on what is needed or what would be better

> * Delegate to lower tier support if the code needs too much effort to review

Reviewer 2: What is lower tier support?

Author: I believe when we wrote this we were thinking about Tier 1,2,3
On a spectrum of community support, technical developer advocate, software engineer, senior software engineer
Each of these can delegate one level down
As an example, if someone submits a very complex commit that eventually will need a senior engineer to review, but the code style or lack of documentation etc. is hindering making it an efficient use of time, then it would get kicked back to DA or a community support person.

Publicly we may not need to mention this as this is more of a concern for senior engineers who belong to organizations that are part of the foundation, and not necessarily a public contributor who wants to volunteer time for code reviews.

Reviewer 2: Sounds good, agreed we can keep it out of the public doc

        * Use the imperative mood in the subject line
        * Wrap the body at 72 characters
        * Use the body to explain what and why vs. how
    * Ensure all code is peer reviewed within your own organization or peers before submitting

>     * Implement and/or address existing feedback before requesting further review

Reviewer 3: Avoid slashes. See [GitLab doc recommended word list entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes) for advice and suggestions.

    * Make a good faith effort to locate example or referential code before requesting someone else direct you towards it
    * If you see a lack of documentation, create a separate commit with draft documentation to fill the gap
        * This documentation can be iterated on same as any code commit, demonstrate in real time that you are learning the code section

>     * Implement or modify relevant tests, otherwise provide clear documentation why test updates were not necessary

Reviewer 3: ```suggestion
    * Implement or modify relevant tests, otherwise provide clear explanation why test updates were not necessary
```
Suggest "explanation" to avoid using "documentation" in a way different than the (user-facing) documentation being discussed in the immediately previous lines, to reduce ambiguity.

    * Tag your PR with affected code areas as best as you can, it’s okay to tag too many, better to cut down irrelevant tags than miss getting input from relevant subject matter experts

- Example PR message
    - please refer [here](https://github.com/prestodb/presto/blob/master/pull_request_template.md)

### What not to do for Pull Requests
* Submit before getting peer review in your own organization
* Request review without addressing/implementing previous feedback
* Asking reviewers to provide examples or code references instead of finding them on your own
* Protesting lack of documentation for a code section

>     * Review the related code, draft initial documentation as a separate commit

Reviewer 3: I don't understand the meaning of the indent for "Review the related code, draft initial documentation as a separate commit". 
Is this another "what not to do", or a recommendation of what you should do instead of "Protesting lack of documentation for a code section"?

Author: The indent is a suggestion on what you should do instead, I will add some language around that

> * Submit without test cases or clear justification for lack there of

Reviewer 3: ```suggestion
* Submit without test cases or clear justification for lack thereof
```

## Code Reviews
#### What to do
* Provide explicit feedback on what is needed or what would be better
* Delegate to lower tier support if the code needs too much effort to review

> * We review code with the objective of helping someone land their changes

Reviewer 4: ```suggestion
* Review code with the objective of helping someone land their changes
```
